### PR TITLE
Typescript definition file

### PR DESCRIPTION
**Pull request by RonNewcomb**
Typescript definition file.  Needs to land in  /node_modules/jQuery-QueryBuilder/  so I guess I put it here?

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+            error?: string;//'glyphicon glyphicon-warning-sign'
+        };
+        templates?: object;
+    }
+
+}
+
+interface JQuery {
+    queryBuilder(method: 'validate', options?: { skip_empty: boolean }): void;
+    queryBuilder(method: 'getRules', options?: { skip_empty: boolean, get_flags: boolean | 'all', allow_invalid: boolean }): jQueryQueryBuilder.Conjunction | jQueryQueryBuilder.Rule[];
```
Comment (mistic100) [inline index.d.ts]: This always returns Conjunction

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        readonly settings: object;
+        readonly templates: Map<string, Function>;
+    }
+
+    interface QueryBuilderEvent extends Event {
+        builder: QueryBuilder;
+    }
+
+    interface TriggerHandler {
+        <T>(e: QueryBuilderEvent, rule: Rule, error: string, value: T): void;
```
Comment (mistic100) [inline index.d.ts]: this is not correct, the values after `e` depends on the trigger

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    }
+
+    interface ChangerHandler {
+        <T>(e: QueryBuilderEvent & { value: T }, level: number): void;
```
Comment (mistic100) [inline index.d.ts]: same

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    }
+
+    interface Conjunction {
+        condition: string;
+        rules: Rule[];
```
Comment (mistic100) [inline index.d.ts]: Actually it's not a Rule object but a plain Json, it should be a new RuleJson type

File: index.d.ts [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        isRoot(): boolean;
+        getPos(): number;
+        drop(): void;
+        moveAfter(ruleOrGroup: Node): void;
+        moveAtBeing(group: Group): void;
+        moveAtEnd(group: Group): void;
+    }
+
+    interface Group extends Node {
+        condition: string;
```
Comment (mistic100) [inline index.d.ts]: missing `readonly rules: Rule[];`

**Comment by RonNewcomb** [comment]
updated my local copy, then. Thanks!

**Comment by mistic100** [comment]
If you did it you could push it anyway....

**Comment by RonNewcomb** [comment]
Done, new PR created.

**Pull request closed without merging**